engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -744,14 +744,17 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
 
     protected <T extends VMInstanceVO> boolean changeState(final T vm, final Event event, final Long hostId, final ItWorkVO work, final Step step) throws NoTransitionException {
         // FIXME: We should do this better.
-        final Step previousStep = work.getStep();
-        _workDao.updateStep(work, step);
+        Step previousStep = null;
+        if (work != null) {
+            previousStep = work.getStep();
```

Comment:
Do you think something like:

    if (!result && work != null) {

would be better?  Even if work.getStep() did return a null, that should have the same effect as before.  Maybe it would be more readable too?